src/include/migraphx/fp4_casts.hpp

```diff
@@ -0,0 +1,166 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2015-2025 Advanced Micro Devices, Inc. All rights reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#ifndef MIGRAPHX_GUARD_RTGLIB_FLOAT4_CASTS_HPP
+#define MIGRAPHX_GUARD_RTGLIB_FLOAT4_CASTS_HPP
+
+#include <cstdint>
+#include <algorithm>
+#include <array>
+#include <migraphx/bit_cast.hpp>
+#include <migraphx/requires.hpp>
+#include <migraphx/errors.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+namespace {
+static std::array<float, 16> _fp4_lut = {
+    0.0, 0.5, 1.0, 1.5, 2.0, 3.0, 4.0, 6.0, -0.0, -0.5, -1.0, -1.5, -2.0, -3.0, -4.0, -6.0};
+}
+
+constexpr float fp4_to_float(uint8_t x) { return _fp4_lut[x & 0xF]; }
+
+// roundTiesToEven
+// based on code in float8_impl
+constexpr uint8_t float_to_fp4(float f_x)
+{
+    // float32 mantissa_width
+    const uint32_t mfmt = 23;
+    // float4 mantissa_width
+    const uint32_t mantissa_size = 1;
+
+    uint32_t x        = migraphx::bit_cast<uint32_t>(f_x);
+    uint32_t head     = 0;
+    uint32_t mantissa = 0;
+    int exponent      = 0;
+    uint32_t bias     = 0;
+    uint32_t sign     = 0;
+    head              = x & 0xFF800000;
+    mantissa          = x & 0x7FFFFF;
+    exponent          = (head >> 23) & 0xFF;
+    sign              = head >> 31;
+    bias              = 127;
```

Comment:
Can you reuse the `get_parts` and `float32_parts` from generic_float.hpp to read the mantissa, exponent, and sign?